lib/ruby_smb/gss/provider.rb

```diff
@@ -0,0 +1,35 @@
+module RubySMB
+  module Gss
+    #
+    # This module provides GSS based authentication.
+    #
+    module Provider
+      # A special constant implying that the authenticated user is anonymous.
+      IDENTITY_ANONYMOUS = :anonymous
+      # The result of a processed GSS request.
+      Result = Struct.new(:buffer, :nt_status, :identity)
+
+      #
+      # The base class for a GSS authentication provider. This class defines a common interface and is not usable as a
+      # provider on its own.
+      #
+      class Base
+        # Create a new, client-specific authenticator instance. This new instance is then able to track the unique state
+        # of a particular client / connection.
+        #
+        # @param [Server::ServerClient] server_client the client instance that this the authenticator will be for
+        def new_authenticator(server_client)
+          raise NotImplementedError
+        end
+
+        #
+        # Whether or not anonymous authentication attempts should be permitted.
+        #
+        attr_accessor :allow_anonymous
+      end
+    end
+  end
+end
+
+require 'ruby_smb/gss/provider/authenticator'
+require 'ruby_smb/gss/provider/ntlm'
```

Comment:
This gives me anxiety 😅  Can we put them on the top of the file?

Reply:
I don't think they can be. When I place them up at the top of the file, I receive the following error:

```
  !: ruby_smb:feat/server12:46:14 ruby_smb git diff
diff --git a/lib/ruby_smb/gss/provider.rb b/lib/ruby_smb/gss/provider.rb
index b68be39..97a1241 100644
--- a/lib/ruby_smb/gss/provider.rb
+++ b/lib/ruby_smb/gss/provider.rb
@@ -1,3 +1,6 @@
+require 'ruby_smb/gss/provider/authenticator'
+require 'ruby_smb/gss/provider/ntlm'
+
 module RubySMB
   module Gss
     #
@@ -30,6 +33,3 @@ module RubySMB
     end
   end
 end
-
-require 'ruby_smb/gss/provider/authenticator'
-require 'ruby_smb/gss/provider/ntlm'
  !: ruby_smb:feat/server12:46:16 ruby_smb ruby examples/auth_capture.rb 
Traceback (most recent call last):
	14: from examples/auth_capture.rb:4:in `<main>'
	13: from examples/auth_capture.rb:4:in `require'
	12: from /home/smcintyre/Repositories/ruby_smb/lib/ruby_smb.rb:13:in `<top (required)>'
	11: from /home/smcintyre/Repositories/ruby_smb/lib/ruby_smb.rb:17:in `<module:RubySMB>'
	10: from /home/smcintyre/Repositories/ruby_smb/lib/ruby_smb.rb:17:in `require'
	 9: from /home/smcintyre/Repositories/ruby_smb/lib/ruby_smb/gss.rb:1:in `<top (required)>'
	 8: from /home/smcintyre/Repositories/ruby_smb/lib/ruby_smb/gss.rb:4:in `<module:RubySMB>'
	 7: from /home/smcintyre/Repositories/ruby_smb/lib/ruby_smb/gss.rb:5:in `<module:Gss>'
	 6: from /home/smcintyre/Repositories/ruby_smb/lib/ruby_smb/gss.rb:5:in `require'
	 5: from /home/smcintyre/Repositories/ruby_smb/lib/ruby_smb/gss/provider.rb:2:in `<top (required)>'
	 4: from /home/smcintyre/Repositories/ruby_smb/lib/ruby_smb/gss/provider.rb:2:in `require'
	 3: from /home/smcintyre/Repositories/ruby_smb/lib/ruby_smb/gss/provider/ntlm.rb:1:in `<top (required)>'
	 2: from /home/smcintyre/Repositories/ruby_smb/lib/ruby_smb/gss/provider/ntlm.rb:2:in `<module:RubySMB>'
	 1: from /home/smcintyre/Repositories/ruby_smb/lib/ruby_smb/gss/provider/ntlm.rb:3:in `<module:Gss>'
/home/smcintyre/Repositories/ruby_smb/lib/ruby_smb/gss/provider/ntlm.rb:8:in `<module:Provider>': uninitialized constant RubySMB::Gss::Provider::Base (NameError)
Did you mean?  Base64
exit status: 1                                                                                                                                                                                                                                                                                                              
  !: ruby_smb:feat/server12:46:22 ruby_smb 
```

Which I'm pretty confident is related to the fact that the `RubySMB::Gss::Provder` module needs to be defined before those files are loaded. There's a similar pattern in the existing [smb1.rb](https://github.com/rapid7/ruby_smb/blob/53e4855b76a63f9ded497182b07a9181b820a925/lib/ruby_smb/smb1.rb#L8) and [smb2.rb](https://github.com/rapid7/ruby_smb/blob/53e4855b76a63f9ded497182b07a9181b820a925/lib/ruby_smb/smb2.rb#L9). In each of those however the requires are within the module definition. I can move them there if that would be better, but I don't think they can be at the top.

Reply:
No that makes sense! Thanks for the explanation 👍